app/controllers/products_controller.rb

```diff
@@ -0,0 +1,73 @@
+class ProductsController < ApplicationController
+skip_before_action :require_login, except: [:index]
+
+  def require_login
+    if current_user.nil?
+      flash[:error] = "You must be logged in to view this section"
+      redirect_to login_path
+    end
+  end
+
+  def index
+    if params[:category_id]
+      @products = Category.find(params[:category_id]).products
```

Comment:
BUG: Submitting the categories form without choosing a category (that is, leaving it on "Search By Category" and clicking Go) creates an error because `params[:category_id]` is nil. You can check for the existence of `params[category_id]` in the controller and add a `flash[:error]` message and rerender the page.